mgr/volumes: Add a per subvolume trash

ShyamsundarR opened the pull request:
Add a per subvolume trash, that is symlinked to the global volume level trash, to avoid renaming subvolume incarnations to the global trash.

This is required as ceph#36472 introduces changes that disallow nested nested snapshots in a subtree (subvolume) and renames across subvolumes. This effect asynchronous purge in mgr/volumes as subvolume are moved to a trash directory for asynchronous deletion by purge threads.

Fixes: https://tracker.ceph.com/issues/47154

Signed-off-by: Venky Shankar <vshankar@redhat.com>
Signed-off-by: Shyamsundar Ranganathan <srangana@redhat.com>

ShyamsundarR commented:
Teuthology run: http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-08-24_23:37:38-fs-wip-shyamsundarr-master-testing-2020-08-24-distro-basic-smithi/

batrick commented:
jenkins test api

vshankar commented:
> Teuthology run: http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-08-24_23:37:38-fs-wip-shyamsundarr-master-testing-2020-08-24-distro-basic-smithi/

looks like this is the entire fs suite. normally for mgr/volumes `--filter volumes` would suffice.

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v2.py:
```diff
@@ -48,6 +49,63 @@ def features(self):
+        # subvolume root is where snapshots would be taken, and hence is the base_path for v2 subvolumes
+        xattr_val = 1
+        try:
+            self.fs.setxattr(self.base_path, 'ceph.dir.subvolume', str(xattr_val).encode('utf-8'), os.XATTR_CREATE)
```
minor nit (can be done later) -- make mark_subvolume() accept path and call that via super() from v2.

batrick commented:
I've queued another: https://pulpito.ceph.com/pdonnell-2020-08-25_02:41:25-fs-wip-shyamsundarr-master-testing-2020-08-24-distro-basic-smithi/

vshankar commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -54,6 +54,19 @@ def path(self):
+        # subvolume root is where snapshots would be taken, and hence is the <uuid> dir for v1 subvolumes
+        xattr_val = 1
+        try:
+            self.fs.setxattr(self.path, 'ceph.dir.subvolume', str(xattr_val).encode('utf-8'), os.XATTR_CREATE)
```
I think XATTR_{CREATE,REPLACE} checks are not available for ceph vxatts -- this would never return EEXIST.
 
So, either we just not bother using this flag or check if the subvolume is already marked in Server::handle_set_vxattr() when handling `ceph.dir.subvolume`:

@ukernel @batrick

batrick replied:
I think removing the flag makes sense. Logically all directories have this "flag" either off or on. So this xattr cannot be created.

vshankar replied:
this gets called when accessing a subvolume. that would lead to disk IO each time -- that's the reason I proposed to include this flag.

batrick replied:
In the `open` method? Perhaps we should just move the `mark_subvolume` call to quota/snapshot creation?

vshankar replied:
I think it is ok (for now) to leave it here: handle_set_vxattr() for ceph.dir.subvolume safely handles subvolume mark on already marked subvolume: https://github.com/ceph/ceph/blob/master/src/mds/Server.cc#L5633

So it doesn't hit the disk every time -- just that it won't return EEXIST.

vshankar reviewed (commented):
couple of minor nits -- rest looks good.

vshankar commented:
test results w/ a minor fix: https://pulpito.ceph.com/vshankar-2020-08-25_08:45:34-fs-wip-per-subvolume-trash-25-08-20-025854-distro-basic-smithi/

There's one failure: https://pulpito.ceph.com/vshankar-2020-08-25_08:45:34-fs-wip-per-subvolume-trash-25-08-20-025854-distro-basic-smithi/5375448/
This is one of those mgr/volumes "hang" issue -- the mgr/volumes command is a subvolume create:

```
2020-08-25T09:37:17.922+0000 7fcbd1342700  0 log_channel(audit) log [DBG] : from='client.7512 -' entity='client.admin' cmd=[{"prefix": "fs subvolume create", "vol_name": "cephfs", "sub_name": "subvolume_0000000000340508", "target": ["mon-mgr", ""]}]: dispatch
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes INFO volumes.module] Starting _cmd_fs_subvolume_create(prefix:fs subvolume create, sub_name:subvolume_0000000000340508, target:['mon-mgr', ''], vol_name:cephfs) < ""
```

The filesystem id changed and mgr/volumes aborts the old filesystem handle (connection pool):

```
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes DEBUG mgr_util] self.fs_id=2, fs_id=3
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes WARNING mgr_util] filesystem id changed for volume 'cephfs', reconnecting...
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes DEBUG mgr_util] self.fs_id=2, fs_id=3
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes INFO mgr_util] aborting connection from cephfs 'cephfs'
2020-08-25T09:37:17.922+0000 7fcbd1b43700  2 client.5172 unmounting (abort)
```

abort calls never returns -- this log message is never seen: https://github.com/ceph/ceph/blob/master/src/pybind/mgr/mgr_util.py#L138

```
zegrep 'abort' ./remote/smithi013/log/ceph-mgr.y.log.gz 
2020-08-25T09:32:24.448+0000 7fcbf96cd200 20 mgr[py] loaded command rbd task add migration abort name=image_spec,type=CephString
2020-08-25T09:37:17.922+0000 7fcbd1b43700  0 [volumes INFO mgr_util] aborting connection from cephfs 'cephfs'
2020-08-25T09:37:17.922+0000 7fcbd1b43700  2 client.5172 unmounting (abort)
```

vshankar commented:
@ShyamsundarR please take a look HEAD in https://github.com/vshankar/ceph/commits/wip-mgr-volumes-per-subvolume-trash ("mgr/volumes: handle idempotent subvolume marks")

kotreshhr commented:
I tested this PR and it looks good.

ShyamsundarR commented:
> @ShyamsundarR please take a look HEAD in https://github.com/vshankar/ceph/commits/wip-mgr-volumes-per-subvolume-trash ("mgr/volumes: handle idempotent subvolume marks")

Added commit from above branch into this PR.

vshankar reviewed (commented):
LGTM (once checks pass)

ShyamsundarR commented:
Added a test case to ensure subvolume is marked with the vxattr ceph.dir.subvolume.

batrick commented:
test looks good Shyam, thanks!

ShyamsundarR commented:
Test link: http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-08-25_17:21:04-fs-wip-shyamsundarr-master-testing-2020-08-25-distro-basic-smithi/

ShyamsundarR commented:
> Test link: http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-08-25_17:21:04-fs-wip-shyamsundarr-master-testing-2020-08-25-distro-basic-smithi/

**Job: 5375936**
Failure due to: https://tracker.ceph.com/issues/46882

**Job: 5375935/37/38**
Failure due to mount path being incorrect(?) in the newly added test `test_subvolume_marked`
```
2020-08-25T17:59:39.903 INFO:tasks.cephfs_test_runner:======================================================================
2020-08-25T17:59:39.904 INFO:tasks.cephfs_test_runner:FAIL: test_subvolume_marked (tasks.cephfs.test_volumes.TestVolumes)
2020-08-25T17:59:39.904 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-08-25T17:59:39.904 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-08-25T17:59:39.904 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-shyamsundarr-master-testing-2020-08-25/qa/tasks/cephfs/test_volumes.py", line 452, in test_subvolume_marked
2020-08-25T17:59:39.905 INFO:tasks.cephfs_test_runner:    os.rename(srcpath, dstpath)
2020-08-25T17:59:39.905 INFO:tasks.cephfs_test_runner:FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/cephtest/mnt.0/volumes/_nogroup/subvolume_0000000000063212/2fb0a5a2-7274-4781-a026-0ed844b0634d' -> '/home/ubuntu/cephtest/mnt.0/volumes/_nogroup/new_subvol_location'
2020-08-25T17:59:39.905 INFO:tasks.cephfs_test_runner:
2020-08-25T17:59:39.906 INFO:tasks.cephfs_test_runner:During handling of the above exception, another exception occurred:
2020-08-25T17:59:39.906 INFO:tasks.cephfs_test_runner:
2020-08-25T17:59:39.906 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-08-25T17:59:39.907 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_ceph_ceph-c_wip-shyamsundarr-master-testing-2020-08-25/qa/tasks/cephfs/test_volumes.py", line 454, in test_subvolume_marked
2020-08-25T17:59:39.907 INFO:tasks.cephfs_test_runner:    self.assertEqual(ose.errno, errno.EXDEV, "invalid error code on renaming subvolume incarnation out of subvolume directory")
2020-08-25T17:59:39.907 INFO:tasks.cephfs_test_runner:AssertionError: 2 != 18 : invalid error code on renaming subvolume incarnation out of subvolume directory
2020-08-25T17:59:39.908 INFO:tasks.cephfs_test_runner:
2020-08-25T17:59:39.908 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-08-25T17:59:39.908 INFO:tasks.cephfs_test_runner:Ran 45 tests in 1547.293s
2020-08-25T17:59:39.909 INFO:tasks.cephfs_test_runner:
2020-08-25T17:59:39.909 INFO:tasks.cephfs_test_runner:FAILED (failures=1)
```

Local (vstart cluster) runs use the following path:

```
2020-08-25 11:20:01,653.653 INFO:__main__:Running ['python3', '-c', '\nimport os\nos.rename("/tmp/tmpueo59nrm/mnt.0/volumes/_nogroup/subvolume_0000000000907968/a1aaf779-7538-46be-8479-0ffafb8ee0ce", "/tmp/tmpueo59nrm/mnt.0/volumes/_nogroup/moved_subvol")\n']
Traceback (most recent call last):
  File "<string>", line 3, in <module>
OSError: [Errno 18] Invalid cross-device link: '/tmp/tmpueo59nrm/mnt.0/volumes/_nogroup/subvolume_0000000000907968/a1aaf779-7538-46be-8479-0ffafb8ee0ce' -> '/tmp/tmpueo59nrm/mnt.0/volumes/_nogroup/moved_subvol'
```

Looking into this...

(remaining jobs failed much earlier, not investigating them at present)

ShyamsundarR commented:
> 
> Looking into this...

Fixed the test case and updated the PR.

Retested the same [build](https://shaman.ceph.com/builds/ceph/wip-shyamsundarr-master-testing-2020-08-25/), using a private [branch](https://github.com/ShyamsundarR/ceph/commits/wip-per-subvol-trash-fix-qa) with the test case change, results are [here](http://pulpito.front.sepia.ceph.com/shyamsundarr-2020-08-25_19:13:31-fs-wip-shyamsundarr-master-testing-2020-08-25-distro-basic-smithi/). (the three failures in the test run are due to https://tracker.ceph.com/issues/46886 )

Tested using:
```
$ CEPH_QA_MAIL="srangana@redhat.com"; MACHINE_NAME=smithi; CEPH_BRANCH=wip-shyamsundarr-master-testing-2020-08-25
$ teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/ShyamsundarR/ceph  -c $CEPH_BRANCH -m $MACHINE_NAME -s fs -k distro -p 75 -e $CEPH_QA_MAIL --suite-branch wip-per-subvol-trash-fix-qa --filter "volumes" -N 3
```

ShyamsundarR commented:
@batrick added a commit to disable subvolume group snapshots. Will wait for build to succeed before starting a test run.

ShyamsundarR commented:
jenkins test docs

ShyamsundarR commented:
Test run with latest PR contents: http://pulpito.front.sepia.ceph.com:80/shyamsundarr-2020-08-26_01:55:49-fs-wip-shyamsundarr-master-testing-2020-08-25-distro-basic-smithi/

vshankar commented:
> Test run with latest PR contents: http://pulpito.front.sepia.ceph.com:80/shyamsundarr-2020-08-26_01:55:49-fs-wip-shyamsundarr-master-testing-2020-08-25-distro-basic-smithi/

There are a couple of failures -- I'll take a look...

batrick commented:
jenkins test docs

batrick commented:
@vshankar 5376091 looks like another instance of the global lock deadlock. Ugh

I have a branch building with #36805 and #36745 : https://shaman.ceph.com/builds/ceph/wip-pdonnell-testing-20200826.032941/

#36745 fixes the upgrade failures.

If you can, please do run tests with that branch.

vshankar commented:
> If you can, please do run tests with that branch.

ack

vshankar commented:
https://pulpito.ceph.com/vshankar-2020-08-26_05:34:12-fs-wip-pdonnell-testing-20200826.032941-distro-basic-smithi/

vshankar commented:
> https://pulpito.ceph.com/vshankar-2020-08-26_05:34:12-fs-wip-pdonnell-testing-20200826.032941-distro-basic-smithi/

Failures are: https://tracker.ceph.com/issues/47140

batrick commented:
jenkins test make check

ShyamsundarR commented:
Test run: http://pulpito.front.sepia.ceph.com:80/shyamsundarr-2020-08-26_15:44:50-fs-wip-shyamsundarr-master-2020-08-26-distro-basic-smithi/

Using branch [wip-shyamsundarr-master-2020-08-26](https://github.com/ceph/ceph-ci/tree/wip-shyamsundarr-master-2020-08-26) from ceph-ci, that includes PRs #36805 and #36745

ShyamsundarR commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/36773/

ShyamsundarR commented:
> Doc render available at http://docs.ceph.com/ceph-prs/36773/

Renders as expected: https://docs.ceph.com/ceph-prs/36773/cephfs/fs-volumes/#fs-subvolume-groups

batrick commented:
> Test run: http://pulpito.front.sepia.ceph.com:80/shyamsundarr-2020-08-26_15:44:50-fs-wip-shyamsundarr-master-2020-08-26-distro-basic-smithi/
> 
> Using branch [wip-shyamsundarr-master-2020-08-26](https://github.com/ceph/ceph-ci/tree/wip-shyamsundarr-master-2020-08-26) from ceph-ci, that includes PRs #36805 and #36745

Failures are: https://tracker.ceph.com/issues/47140

Good to go!

batrick commented:
@ShyamsundarR added a tracker ticket to the commit messages. Previous jenkins tests passed. Merging.
